### PR TITLE
#161996678 Add a user sign up and database storage.

### PR DESCRIPTION
**Pull request by BrianSerem**
#### What does this PR do?
Add a user sign up through storage of info in an external database.

#### Description of Task to be completed?
Have the following endpoint working with storage of data in a database.
/api/v1/auth/signup

#### How should this be manually tested?
Clone this repository.
Cd into the directory.
Export run.py as a flask application
flask run and use the given url to test the endpoint.

#### Any background context you want to provide?
Every storage of data is now on a database.
After a successful sign up, the user is authenticated and given a unique access token for accessing restricted areas of the app.
 
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/161996678

#### Screenshots
userregistration

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -3,14 +3,17 @@
 from .api.V1 import version_1
 from .api.V2 import version_2
 from app import db_config
+from flask_jwt_extended import JWTManager
 
 
 def create_app():
     """Crete app fuction."""
     app = Flask(__name__)
+    app.secret_key = "brian"
```
Comment (cosmas28) [inline app/__init__.py]: This secret key should be moved to dotenv file. This is a good practice.

File: app/api/V2/models/usermodels.py [hunk 2]
```diff
@@ -1,36 +1,57 @@
 """COntain User model."""
```
Comment (cosmas28) [inline app/api/V2/models/usermodels.py]: kindly fix the typos.

File: app/api/V2/__init__.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 from .views.parcelviews import MovingParcels, SpecificParcel, DeliverParcel
 from .views.parcelviews import DeclinedParcels, MarkParcelInTransit
 from .views.parcelviews import ApproveParcel, GetAcceptedParcels, DeclineParcel
-from .views.parcelviews import UserRegistration
+from .views.userviews import UserRegistration
 
 version_2 = Blueprint('v2', __name__)
```
Comment (cosmas28) [inline app/api/V2/__init__.py:13]: there should be only one line space at the end of the files. Let this be the last time I'm repeating this feedback.

File: app/api/V2/models/usermodels.py [hunk 4]
```diff
@@ -1,36 +1,57 @@
...
+        row = cur.fetchone()
+        if row:
+            user = cls(*row)
+        else:
+            user = None
+
+        return user
+
+
+
```
Comment (cosmas28) [inline app/api/V2/models/usermodels.py]: Too many lines at the end of this file. Kindly leave one line only.

File: app/api/V2/models/usermodels.py [hunk 5]
```diff
@@ -1,36 +1,57 @@
...
         )
 
-    def get_by_username(self, username):
+    @classmethod
+    def get_by_username(cls, username):
         """Find a user by username."""
-        for user in users:
-            if user.username == username:
-                return user
+        cur = connection.cursor()
```
Comment (cosmas28) [inline app/api/V2/models/usermodels.py]: Kindly use a more descriptive variable name than `cur`.

File: app/api/V2/models/usermodels.py [hunk 6]
```diff
@@ -1,36 +1,57 @@
...
             username=self.username,
             email=self.email,
-            password=self.password
+            role=self.role,
+            phone=self.phone
         )
 
-    def get_by_username(self, username):
+    @classmethod
+    def get_by_username(cls, username):
```
Comment (cosmas28) [inline app/api/V2/models/usermodels.py]: This function is not descriptive enough. I think it should be `get_user_by_username`.

File: app/api/V2/views/userviews.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            cur = connection.cursor() 
+            cur.execute(query, (data['username'], data['email'],data['password'],data['phone'],data['role']))
+            connection.commit()
+            user = User().get_by_username(data['username'])
+            access_token = create_access_token(identity=user.id, fresh=True)
+            refresh_token = create_refresh_token(user.id)
+            return {"access token":access_token}, 200
+            return {"error":"error connecting to the database"}, 404
+
+        return{"error":"username {} is already taken".format(data['username'])}
```
Comment (cosmas28) [inline app/api/V2/views/userviews.py]: Leave a line space at the end of each file.

File: app/api/V2/views/userviews.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        user = User().get_by_username(data['username'])
+        if not user:
+            cur = connection.cursor() 
+            cur.execute(query, (data['username'], data['email'],data['password'],data['phone'],data['role']))
+            connection.commit()
+            user = User().get_by_username(data['username'])
+            access_token = create_access_token(identity=user.id, fresh=True)
+            refresh_token = create_refresh_token(user.id)
+            return {"access token":access_token}, 200
+            return {"error":"error connecting to the database"}, 404
```
Comment (cosmas28) [inline app/api/V2/views/userviews.py]: You can only have one return statement in a block of code like `if`. The second return statement can't work.

File: app/__init__.py (hunk 1 above)
Comment (BrianSerem) [reply app/__init__.py]: I created a variable secret_key inside the .env file

File: app/api/V2/models/usermodels.py (hunk 2 above)
Comment (BrianSerem) [reply app/api/V2/models/usermodels.py]: Fixed the typo in the usermodels.py  header doctsring

File: app/api/V2/__init__.py (hunk 3 above)
Comment (BrianSerem) [reply app/api/V2/__init__.py:13]: Deleted the extra lines at the end of the file

File: app/api/V2/models/usermodels.py (hunk 4 above)
Comment (BrianSerem) [reply app/api/V2/models/usermodels.py]: removed the extra pieces of lines at the end of the file

File: app/api/V2/models/usermodels.py (hunk 5 above)
Comment (BrianSerem) [reply app/api/V2/models/usermodels.py]: I have used cursorobject for the previously "cur" variable

File: app/api/V2/models/usermodels.py (hunk 6 above)
Comment (BrianSerem) [reply app/api/V2/models/usermodels.py]: I changed its name to get_user_by_username as you suggested

File: app/api/V2/views/userviews.py (hunk 7 above)
Comment (BrianSerem) [reply app/api/V2/views/userviews.py]: Left a single a the end of all files in this PR

File: app/api/V2/views/userviews.py (hunk 8 above)
Comment (BrianSerem) [reply app/api/V2/views/userviews.py]: Removed the extra return statement and placed it inside a try block.

**Pull request merged**